### PR TITLE
feat: add tx_template

### PR DESCRIPTION
**Pull request by thevaibhav-dixit**

### CODE REVIEW

File: cala-ledger-core-types/src/tx_template.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    pub layer: CelExpression,
+    pub direction: CelExpression,
+    pub units: CelExpression,
+    pub currency: CelExpression,
+    pub description: Option<CelExpression>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct TxInput {
+    pub effective: String,
```
Comment (bodymindarts) [inline cala-ledger-core-types/src/tx_template.rs:38]: Aren't these `CelExpression`s?

File: cala-ledger-core-types/Cargo.toml [hunk 2]
```diff
@@ -6,9 +6,11 @@ edition = "2021"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
+cala-cel-interpreter = { path = "../cala-cel-interpreter/", package = "cala-cel-interpreter" }
```
Comment (bodymindarts) [inline cala-ledger-core-types/Cargo.toml:9]: Style: I like a line break between the local and external deps.

File: cala-ledger-core-types/src/tx_template.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    INTEGER,
+    DECIMAL,
+    BOOLEAN,
+    UUID,
+    DATE,
+    TIMESTAMP,
+    JSON,
+}
+
+// need a place to handle this
```
Comment (bodymindarts) [inline cala-ledger-core-types/src/tx_template.rs:58]: This is fine - remove comment.

File: cala-ledger/src/tx_template/cel_context.rs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+use cala_cel_interpreter::CelContext;
+
+pub(super) fn initialize() -> CelContext {
```
Comment (bodymindarts) [inline cala-ledger/src/tx_template/cel_context.rs:3]: I don't like the location of this - its okay for now (as its the same as in sqlx-ledger) but the interpreter will need a bit of refactoring anyhow.

File: cala-ledger/src/tx_template/entity.rs [hunk 5]
```diff
@@ -0,0 +1,364 @@
+use derive_builder::Builder;
```
Comment (bodymindarts) [inline cala-ledger/src/tx_template/entity.rs:1]: did you commit the file above `cala-ledger/src/tx_template/core.rs` by accident?

File: cala-ledger/src/tx_template/mod.rs [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    pub async fn create(
+        &self,
+        new_tx_template: NewTxTemplate,
+    ) -> Result<TxTemplate, TxTemplateError> {
+        let mut tx = self.pool.begin().await?;
+        let EntityUpdate {
+            entity: tx_template,
+            ..
+        } = self.repo.create_in_tx(&mut tx, new_tx_template).await?;
+        // self.outbox
```
Comment (bodymindarts) [inline cala-ledger/src/tx_template/mod.rs:42]: why is this commented?

**Review by bodymindarts — APPROVED** [review]
Okay looks good for now!

Missing is outbox + wiring to nodejs and GQL I guess.
Can do it incrementally - or continue with the next entity. Up to you! At least outbox should be added for completeness before moving on though.

File: cala-ledger/src/tx_template/mod.rs (hunk 6 above)
Comment (thevaibhav-dixit) [reply cala-ledger/src/tx_template/mod.rs:42]: commented to make the compiler pass checks. Will work on adding the outbox in the next pr

File: cala-ledger-core-types/src/tx_template.rs (hunk 1 above)
Comment (thevaibhav-dixit) [reply cala-ledger-core-types/src/tx_template.rs:38]: Yeah you are right... will need to change this. thanks !

Comment (thevaibhav-dixit) [reply cala-ledger-core-types/src/tx_template.rs:38]: Yeah you are right... will need to change this. thanks !

File: cala-ledger-core-types/src/tx_template.rs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    pub description: Option<String>,
+    pub metadata: Option<serde_json::Value>,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct ParamDefinition {
+    pub name: String,
+    pub r#type: ParamDataType,
+    pub default: Option<CelExpression>,
+    pub description: Option<String>,
```
Comment (thevaibhav-dixit) [inline cala-ledger-core-types/src/tx_template.rs:22]: In sqlx-ledger ```default``` is  ```Option<String>``` and we have a ```to_default_expr``` that converts ```Option<String>``` to Option<CelExpression```. Should I change it ````Option<String>``` or this works ?

Comment (bodymindarts) [reply cala-ledger-core-types/src/tx_template.rs:22]: this type should have `Option<CelExpression>`. Its the builder that should be `Option<String>`

File: cala-ledger-core-types/Cargo.toml [hunk 8]
```diff
@@ -6,9 +6,12 @@ edition = "2021"
...
```
Comment (bodymindarts) [inline cala-ledger-core-types/Cargo.toml:9]: I don't think you need `package = ...` if it's the same. The only reason for `package = ` is if you want to rename it internal to this crate.

Like probably you could simply call it `cel_interpreter` internally....
(check how its done in sqlx-ledger)

**Review by bodymindarts — APPROVED** [review]
Great - but needs rebasing

**Pull request merged**